src/bicep-types/src/types.ts

```diff
@@ -341,16 +341,55 @@ export class TypeFactory {
     return this.addType(this.booleanType);
   }
 
-  public addResourceType(name: string, scopeType: ScopeType, readOnlyScopes: ScopeType | undefined, body: TypeReference, flags: ResourceFlags, functions?: Record<string, ResourceTypeFunction>) {
-    return this.addType({
+  public addResourceType(
+    name: string,
+    body: TypeReference,
+    scopeType?: ScopeType,
+    readOnlyScopes?: ScopeType,
+    flags?: ResourceFlags,
+    functions?: Record<string, ResourceTypeFunction>,
+    readableScopes?: ScopeType,
+    writableScopes?: ScopeType,
+  ) {
+
+    // Guard against mixing legacy & modern
+    const hasLegacy =
+      scopeType !== undefined ||
+      readOnlyScopes !== undefined ||
+      (flags !== undefined && flags !== ResourceFlags.None);
+
+    const hasModern =
+      writableScopes !== undefined ||
+      readableScopes !== undefined;
+
+    if (hasLegacy && hasModern) {
+      throw new Error(
+        "Cannot mix legacy scope fields (scopeType, readOnlyScopes, flags) " +
+        "with modern fields (writableScopes, readableScopes).");
+    }
```

Comment:
I don't think you need to worry about backwards compatibility in the JS code. Given how this repo is consumed, the JS is for _writing_ type data, and the C# is for _reading_ type data. So realistically, you can make some pretty dramatic breaking changes to the JS since the only consumers are the type generators, and since scopes are really only relevant to ARM resources, the only consumer that'll need to be updated for this change is `bicep-types-az`.